x/cronos/keeper/precompiles/ica.go

```diff
@@ -0,0 +1,218 @@
+	SubmitMsgsMethod      abi.Method
+	IcaContractAddress    = common.BytesToAddress([]byte{102})
+)
+
+func init() {
+	addressType, _ := abi.NewType("address", "", nil)
+	stringType, _ := abi.NewType("string", "", nil)
+	bytesType, _ := abi.NewType("bytes", "", nil)
+	uint256Type, _ := abi.NewType("uint256", "", nil)
+	RegisterAccountMethod = abi.NewMethod(
+		"registerAccount", "registerAccount", abi.Function, "", false, false, abi.Arguments{abi.Argument{
+			Name: "connectionID",
+			Type: stringType,
+		}, abi.Argument{
+			Name: "owner",
+			Type: addressType,
+		}, abi.Argument{
+			Name: "version",
+			Type: stringType,
+		}},
+		abi.Arguments{abi.Argument{
+			Name: "res",
+			Type: bytesType,
+		}},
+	)
+	QueryAccountMethod = abi.NewMethod(
+		"queryAccount", "queryAccount", abi.Function, "", false, false, abi.Arguments{abi.Argument{
+			Name: "connectionID",
+			Type: stringType,
+		}, abi.Argument{
+			Name: "owner",
+			Type: addressType,
+		}},
+		abi.Arguments{abi.Argument{
+			Name: "res",
+			Type: bytesType,
+		}},
+	)
+	SubmitMsgsMethod = abi.NewMethod(
+		"submitMsgs", "submitMsgs", abi.Function, "", false, false, abi.Arguments{abi.Argument{
+			Name: "connectionID",
+			Type: stringType,
+		}, abi.Argument{
+			Name: "owner",
+			Type: addressType,
+		}, abi.Argument{
+			Name: "data",
+			Type: stringType,
+		}, abi.Argument{
+			Name: "timeout",
+			Type: uint256Type,
+		}},
+		abi.Arguments{abi.Argument{
+			Name: "res",
+			Type: bytesType,
+		}},
+	)
+}
+
+type IcaContract struct {
+	BaseContract
+
+	cdc           codec.Codec
+	icaauthKeeper *icaauthkeeper.Keeper
+}
+
+func NewIcaContract(icaauthKeeper *icaauthkeeper.Keeper, cdc codec.Codec) vm.PrecompiledContract {
+	return &IcaContract{
+		BaseContract:  NewBaseContract(IcaContractAddress),
+		cdc:           cdc,
+		icaauthKeeper: icaauthKeeper,
+	}
+}
+
+func (ic *IcaContract) Address() common.Address {
+	return IcaContractAddress
+}
+
+// RequiredGas calculates the contract gas use
+func (ic *IcaContract) RequiredGas(input []byte) uint64 {
+	return ICAContractRequiredGas
+}
+
+func (ic *IcaContract) IsStateful() bool {
+	return true
+}
+
+func (ic *IcaContract) Run(evm *vm.EVM, contract *vm.Contract, readonly bool) ([]byte, error) {
+	// parse input
+	methodID := contract.Input[:4]
+	stateDB := evm.StateDB.(ExtStateDB)
+	precompileAddr := ic.Address()
+	converter := cronosevents.IcaConvertEvent
+	var execErr error
+	var res codec.ProtoMarshaler
+	switch string(methodID) {
+	case string(RegisterAccountMethod.ID):
+		if readonly {
+			return nil, errors.New("the method is not readonly")
+		}
+		args, err := RegisterAccountMethod.Inputs.Unpack(contract.Input[4:])
+		if err != nil {
+			return nil, errors.New("fail to unpack input arguments")
+		}
+		connectionID := args[0].(string)
+		account := args[1].(common.Address)
+		version := args[2].(string)
+		txSender := evm.Origin
+		if !isSameAddress(account, contract.CallerAddress) && !isSameAddress(account, txSender) {
```

Comment:
we shouldn't check for `txSender`, because the tx sender may not want to authorize the contract to operate on his behalf.